### PR TITLE
Per-Key Data Limits in Shadowbox

### PR DESCRIPTION
**Pull request by JonathanDCohen**
This PR adds the ability to create data limits on individual keys in shadowbox.  This includes two new API endpoints, `{PUT, DELETE} /access-keys/$id/data-limit {limit: {bytes: Number}}`.  The number of keys with data limits set on them is sent to the feature metrics table if the server is opted in to metrics sharing.

As part of this request, language around "AccessKeyDataLimit" which refers to the server default data limit was changed to use language like "DefaultDataLimit", since AccessKeyDataLimit makes more sense as the limit for an individual access key.

Tested:
  shadowbox unit and integration tests

  metrics server unit and integration tests

  metrics manual integration test:  Set up a local shadowbox server and created an access key with a data limit.  Verified that a row showing the data limit showed up in the dev BigQuery feature metrics table

  Data limits manual integration test: Set up a shadowbox server with two keys -- one with a low individual data limit and one with a high server-default data limit.  I ran a speedtest on the first key to exhaust its limit and verified that I was unable to use the key afterwards to use the Internet.  I then switched to the other key and attended a team meeting through the proxy to verify that it was still working.

### CODE REVIEW

File: src/metrics_server/feature_metrics.ts [hunk 1]
```diff
@@ -66,10 +66,22 @@ export function isValidFeatureMetricsReport(obj: any): obj is DailyFeatureMetric
...
 
-  // Validate individual feature records.
+  // Validate the server data limit feature
   if (typeof obj.dataLimit.enabled !== 'boolean') {
     return false;
   }
 
+  // Validate the per-key data limit feature
+  const perKeyType = typeof obj.dataLimit.perKeyLimitCount;
+  if (perKeyType !== 'undefined') {
```
Comment (alalamav) [inline src/metrics_server/feature_metrics.ts:76]: nit: may be clearer to check `obj.dataLimit.perKeyLimitCount === undefined`.

File: src/shadowbox/server/manager_service.ts [hunk 2]
```diff
@@ -125,6 +128,19 @@ function validateAccessKeyId(accessKeyId: unknown): string {
   return accessKeyId;
 }
 
+function validateDataLimit(limit: unknown): DataLimit {
```
Comment (alalamav) [inline src/shadowbox/server/manager_service.ts:131]: You could use a [type-guard](https://basarat.gitbook.io/typescript/type-system/typeguard#user-defined-type-guards) avoid reasigning the returned value:

```typescript
function isDataLimit(limit: unknown): is DataLimit {
```

File: src/shadowbox/server/manager_service.ts [hunk 3]
```diff
@@ -28,7 +28,7 @@ import {ServerConfigJson} from './server_config';
 import {SharedMetricsPublisher} from './shared_metrics';
 
 // Creates a AccessKey response.
-function accessKeyToJson(accessKey: AccessKey) {
+function accessKeyResponseJson(accessKey: AccessKey) {
```
Comment (alalamav) [inline src/shadowbox/server/manager_service.ts:31]: nit: I think the previous name was clearer as it implied an action: convert/serialize.

File: src/shadowbox/server/server_access_key.ts [hunk 4]
```diff
@@ -70,14 +67,15 @@ function makeAccessKey(hostname: string, accessKeyJson: AccessKeyJson): AccessKe
       accessKeyJson.id, accessKeyJson.name, accessKeyJson.metricsId, proxyParams);
 }
 
-function makeAccessKeyJson(accessKey: AccessKey): AccessKeyJson {
+function accessKeySerializedJson(accessKey: AccessKey): AccessKeyJson {
```
Comment (alalamav) [inline src/shadowbox/server/server_access_key.ts:70]: nit: I think the previous name was clearer since it denotes an action

File: src/shadowbox/server/server_access_key.ts [hunk 5]
```diff
@@ -47,11 +48,7 @@ class ServerAccessKey implements AccessKey {
   public isOverDataLimit = false;
   constructor(
       readonly id: AccessKeyId, public name: string, public metricsId: AccessKeyMetricsId,
-      readonly proxyParams: ProxyParams) {}
-}
-
-function isValidAccessKeyDataLimit(limit: DataLimit): boolean {
```
Comment (alalamav) [inline src/shadowbox/server/server_access_key.ts:53]: Why did you remove validation in this class? Could be useful for unit testing.

File: src/shadowbox/server/server_access_key.ts [hunk 6]
```diff
@@ -212,8 +219,7 @@ export class ServerAccessKeyRepository implements AccessKeyRepository {
     for (const accessKey of this.accessKeys) {
       const usageBytes = bytesTransferredById[accessKey.id] || 0;
       const wasOverDataLimit = accessKey.isOverDataLimit;
-      accessKey.isOverDataLimit =
-          this.accessKeyDataLimit ? usageBytes > this.accessKeyDataLimit.bytes : false;
+      accessKey.isOverDataLimit = usageBytes > (accessKey.dataLimit || this.defaultDataLimit)?.bytes || false;
```
Comment (alalamav) [inline src/shadowbox/server/server_access_key.ts:222]: Can you extract the limit to a separate variable and keep the ternary operator to avoid `|| false`?

File: src/shadowbox/server/manager_service.ts [hunk 7]
```diff
@@ -302,32 +318,58 @@ export class ShadowsocksManagerService {
...
-      const limit = req.params.limit as DataLimit;
-      if (!limit) {
-        return next(new restifyErrors.MissingParameterError(
-            {statusCode: 400}, 'Missing `limit` parameter'));
-      } else if (!Number.isInteger(limit.bytes)) {
-        return next(new restifyErrors.InvalidArgumentError(
-            {statusCode: 400}, '`limit` must be an integer'));
+      const accessKeyId = validateAccessKeyId(req.params.id);
+      const limit = validateDataLimit(req.params.limit);
+      await this.accessKeys.setAccessKeyDataLimit(accessKeyId, limit);
```
Comment (alalamav) [inline src/shadowbox/server/manager_service.ts:323]: Are you awaiting for enforcement to happen? It may be more efficient not to wait - you already know that settting the limit succeeded synchronously.

File: src/shadowbox/server/manager_service.ts [hunk 8]
```diff
@@ -302,32 +318,58 @@ export class ShadowsocksManagerService {
...
+      }
+      return next(error);
+    }
+  }
+
+  public async removeAccessKeyDataLimit(req: RequestType, res: ResponseType, next: restify.Next) {
+    try {
+      logging.debug(`removeAccessKeyDataLimit request ${JSON.stringify(req.params)}`);
+      const accessKeyId = validateAccessKeyId(req.params.id);
+      await this.accessKeys.removeAccessKeyDataLimit(accessKeyId);
```
Comment (alalamav) [inline src/shadowbox/server/manager_service.ts:339]: Same here.

File: src/shadowbox/model/access_key.ts [hunk 9]
```diff
@@ -62,7 +64,11 @@ export interface AccessKeyRepository {
...
   // Sets a data transfer limit for all access keys.
-  setAccessKeyDataLimit(limit: DataLimit): Promise<void>;
+  setDefaultDataLimit(limit: DataLimit): Promise<void>;
   // Removes the access key data transfer limit.
-  removeAccessKeyDataLimit(): Promise<void>;
+  removeDefaultDataLimit(): Promise<void>;
+  // Sets access key `id` to use the given custom data limit.
+  setAccessKeyDataLimit(id: AccessKeyId, limit: DataLimit);
+  // Removes the custom data limit from access key `id`.
+  removeAccessKeyDataLimit(id: AccessKeyId);
```
Comment (alalamav) [inline src/shadowbox/model/access_key.ts:73]: Your implementations of `[set|remove]AccessKeyDataLimit` return `Promise<void>`, let's update the interface to reflect that.

File: src/shadowbox/server/shared_metrics.ts [hunk 10]
```diff
@@ -233,11 +236,15 @@ export class OutlineSharedMetricsPublisher implements SharedMetricsPublisher {
...
   private async reportFeatureMetrics(): Promise<void> {
+    const keys = this.keyConfig.data().accessKeys;
     const featureMetricsReport = {
       serverId: this.serverConfig.data().serverId,
       serverVersion: version,
       timestampUtcMs: this.clock.now(),
-      dataLimit: {enabled: !!this.serverConfig.data().accessKeyDataLimit},
+      dataLimit: {
+        enabled: !!this.serverConfig.data().accessKeyDataLimit,
+        perKeyLimitCount: keys?.reduce((count, next) => count + (next.dataLimit ? 1 : 0), 0) || 0
```
Comment (alalamav) [inline src/shadowbox/server/shared_metrics.ts:246]: nit: may be more readable to extract the counting to a separate function.

File: src/shadowbox/server/shared_metrics.ts [hunk 11]
```diff
@@ -156,6 +158,7 @@ export class OutlineSharedMetricsPublisher implements SharedMetricsPublisher {
   // metricsUrl: where to post the metrics
   constructor(
       private clock: Clock, private serverConfig: JsonConfig<ServerConfigJson>,
+      private keyConfig: JsonConfig<AccessKeyConfigJson>,
```
Comment (alalamav) [inline src/shadowbox/server/shared_metrics.ts:161]: Why not pass the access key repository instead of the backing file?

**Review by alalamav — COMMENTED** [review]
This is looking good! I have some suggestions in the implementation; haven't reviewed the tests in detail yet.

File: src/shadowbox/server/manager_service.ts (hunk 2 above)
Comment (JonathanDCohen) [reply src/shadowbox/server/manager_service.ts:131]: I didn't use a type guard because the function would always either throw or return true.  I use exceptions here since this function knows what the exact error this is, allowing us to return consistent errors on ever endpoint which takes a user-supplied data limit. 

Using a type guard would set up a situation where we have

```
if isDataLimit(req.params.limit) {
  // The entire rest of the function which uses the limit
}
// some other code which looks like it will run if the limit isn't valid, but actually won't ever be run
// unless isDataLimit is true because otherwise we'll throw and enter the exception handler
```

This looks really weird at the call-site to me.

File: src/shadowbox/server/server_access_key.ts (hunk 4 above)
Comment (JonathanDCohen) [inline src/shadowbox/server/server_access_key.ts:70]: In your previous comment you noted taht both of these functions serialize, which is a good point.  I think accessKeyToPersistedJson satisfies both of our concerns.

File: src/shadowbox/server/server_access_key.ts (hunk 5 above)
Comment (JonathanDCohen) [reply src/shadowbox/server/server_access_key.ts:53]: I removed it since this was duplicating the validation done at the API boundary.  We can export the validation function from manager_service if we need for testing, but according to the layering we've set up we're already in typescript here with a known-valid data limit, so it seemed redundant to validate again.

We weren't using it for unit testing anyways.

File: src/shadowbox/server/shared_metrics.ts (hunk 10 above)
Comment (JonathanDCohen) [reply src/shadowbox/server/shared_metrics.ts:246]: Done

File: src/shadowbox/server/shared_metrics.ts (hunk 11 above)
Comment (JonathanDCohen) [reply src/shadowbox/server/shared_metrics.ts:161]: A couple reasons:

First, the access key repo has no way to get a single key without copying the whole list of keys without adding functionality.  Passing the config allows us to directly access the keys, and the config class is even written with the metrics use case in mind.  

Second, we don't need any of the other functionality of the access key repo than just listing the keys, which we get for free with the key config.

Third, stylistic parity with the act that we already pass the server config.

File: src/metrics_server/feature_metrics.ts (hunk 1 above)
Comment (JonathanDCohen) [reply src/metrics_server/feature_metrics.ts:76]: Done.  I restructured the branching logic as well and I think it's a whole lot clearer now, so thanks for the prod!

File: src/shadowbox/server/manager_service.ts (hunk 2 above)
Comment (JonathanDCohen) [reply src/shadowbox/server/manager_service.ts:131]: I didn't use a type guard because the function would always either throw or return true.  I use exceptions here since this function knows what the exact error this is, allowing us to return consistent errors on ever endpoint which takes a user-supplied data limit. 

Using a type guard would set up a situation where we have

```
if isDataLimit(req.params.limit) {
  // The entire rest of the function which uses the limit
}
// some other code which looks like it will run if the limit isn't valid, but actually won't ever be run
// unless isDataLimit is true because otherwise we'll throw and enter the exception handler
```

This looks really weird at the call-site to me.

File: src/shadowbox/server/server_access_key.ts (hunk 6 above)
Comment (JonathanDCohen) [reply src/shadowbox/server/server_access_key.ts:222]: Done

Comment (JonathanDCohen) [reply src/shadowbox/server/server_access_key.ts:222]: The ternary has a small bug that it skips the check and returns false if the data limit is 0, since 0 is falsy.

I extracted to a separate variable for readability.

File: src/shadowbox/server/server_access_key.ts (hunk 4 above)
Comment (JonathanDCohen) [reply src/shadowbox/server/server_access_key.ts]: Done in  c9ae52d

File: src/shadowbox/server/manager_service.ts (hunk 3 above)
Comment (JonathanDCohen) [reply src/shadowbox/server/manager_service.ts]: Done on  c9ae52d

**Comment by JonathanDCohen** [comment]
Sorry for the mess of responses. I've been using a GitHub code review extension in VSCode and they just had an update.  I'm not sure how some of my comments ended up both as review responses as well as replies to your comments.

File: src/shadowbox/server/manager_service.ts (hunk 2 above)
Comment (alalamav) [reply src/shadowbox/server/manager_service.ts:131]: Fair enough, I missed that you were throwing specific errors.

File: src/shadowbox/server/manager_service.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (alalamav) [reply src/shadowbox/server/manager_service.ts:323]: We actually don't wait for default limits - perhaps @fortuna can weigh in.

File: src/shadowbox/server/server_access_key.spec.ts [hunk 12]
```diff
@@ -220,22 +302,22 @@ describe('ServerAccessKeyRepository', () => {
...
+    const repo = new RepoBuilder().defaultDataLimit(limit).build();
     expect(repo.dataLimit).toEqual(limit);
-    await expectNoAsyncThrow(repo.removeAccessKeyDataLimit.bind(repo));
+    await expectNoAsyncThrow(repo.removeDefaultDataLimit.bind(repo));
     expect(repo.dataLimit).toBeUndefined();
     done();
   });
 
-  it('removeAccessKeyDataLimit restores over-limit access keys', async (done) => {
+  it('removeDefaultDataLimit restores over-limit access keys', async (done) => {
```
Comment (alalamav) [inline src/shadowbox/server/server_access_key.spec.ts:314]: I think it may be useful to enhance the limit enforcement test to make sure we are enforcing per-key limits.

File: src/shadowbox/server/manager_service.ts [hunk 13]
```diff
@@ -28,7 +28,7 @@ import {ServerConfigJson} from './server_config';
 import {SharedMetricsPublisher} from './shared_metrics';
 
 // Creates a AccessKey response.
-function accessKeyToJson(accessKey: AccessKey) {
+function accessKeyToResponse(accessKey: AccessKey) {
```
Comment (alalamav) [inline src/shadowbox/server/manager_service.ts:31]: nit: JsonResponse?

File: src/shadowbox/server/server_access_key.ts [hunk 14]
```diff
@@ -12,6 +12,7 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+import { access } from 'fs';
```
Comment (alalamav) [inline src/shadowbox/server/server_access_key.ts:15]: Do you need this? I couldn't figure out whether it's used.

File: src/shadowbox/server/server_access_key.ts [hunk 15]
```diff
@@ -212,8 +220,8 @@ export class ServerAccessKeyRepository implements AccessKeyRepository {
     for (const accessKey of this.accessKeys) {
       const usageBytes = bytesTransferredById[accessKey.id] || 0;
       const wasOverDataLimit = accessKey.isOverDataLimit;
-      accessKey.isOverDataLimit =
-          this.accessKeyDataLimit ? usageBytes > this.accessKeyDataLimit.bytes : false;
+      const limitBytes = (accessKey.dataLimit || this.defaultDataLimit)?.bytes;
+      accessKey.isOverDataLimit = usageBytes > limitBytes || false;
```
Comment (alalamav) [inline src/shadowbox/server/server_access_key.ts:224]: Good catch on the 0 data limit bug. I don't think you need the `|| false`, since `number > undefined || undefined > undefined == false`.

File: src/shadowbox/server/manager_service.ts (hunk 13 above)
Comment (JonathanDCohen) [reply src/shadowbox/server/manager_service.ts]: Done in 20ad19e

File: src/shadowbox/server/server_access_key.spec.ts (hunk 12 above)
Comment (JonathanDCohen) [reply src/shadowbox/server/server_access_key.spec.ts:314]: Done in 20ad19e

File: docker [hunk 16]
```diff
@@ -0,0 +1 @@
+{"a": 1} exec -i vibrant_satoshi jq
```
Comment (alalamav) [inline docker:1]: What is this?

File: src/shadowbox/integration_test/util/docker [hunk 17]
```diff
@@ -0,0 +1 @@
+echo hi run util:0.1
```
Comment (alalamav) [inline src/shadowbox/integration_test/util/docker:1]: Same here, do we need it?

File: src/shadowbox/server/manager_service.ts [hunk 18]
```diff
@@ -12,6 +12,7 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+import { access } from 'fs';
```
Comment (alalamav) [inline src/shadowbox/server/manager_service.ts:15]: Remove.

File: src/shadowbox/server/manager_service.ts [hunk 19]
```diff
@@ -38,12 +39,13 @@ function accessKeyToJson(accessKey: AccessKey) {
...
     port: accessKey.proxyParams.portNumber,
     method: accessKey.proxyParams.encryptionMethod,
+    dataLimit: accessKey.dataLimit,
     accessUrl: SIP002_URI.stringify(makeConfig({
       host: accessKey.proxyParams.hostname,
       port: accessKey.proxyParams.portNumber,
       method: accessKey.proxyParams.encryptionMethod,
       password: accessKey.proxyParams.password,
-      outline: 1,
+      outline: 1
```
Comment (alalamav) [inline src/shadowbox/server/manager_service.ts:48]: nit: dangling comma

**Review by alalamav — COMMENTED** [review]
Good catch on the integration test and serialization issues. I have some minor comments, this is looking good.

File: docker (hunk 16 above)
Comment (JonathanDCohen) [reply docker:1]: A missed artifact from debugging the integration test.  Revmoved

File: src/shadowbox/integration_test/util/docker (hunk 17 above)
Comment (JonathanDCohen) [reply src/shadowbox/integration_test/util/docker:1]: Nope, done

File: src/shadowbox/server/shared_metrics.ts (hunk 11 above)
Comment (fortuna) [reply src/shadowbox/server/shared_metrics.ts:162]: Let's not expose the serialization format. The server config was being passed because that was a configuration for the publisher. It's a different case.

Let's pass the repo instead as Alberto suggested. Why do you need to get a single key? Don't you need to go over all of them anyway?

Comment (fortuna) [reply src/shadowbox/server/shared_metrics.ts:162]: Or perhaps the feature metrics could be added to UsageMetrics. Or create a FeatureMetrics object like that one.

File: src/shadowbox/server/manager_service.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (fortuna) [reply src/shadowbox/server/manager_service.ts:324]: I agree with Jon that we should wait for the config to be written, but we should not wait for enforcement. Doesn't that happen out of band anyway?

**Review by fortuna — COMMENTED** [review]
I'm excited to see this coming to life!!!

File: src/shadowbox/server/server_access_key.ts [hunk 20]
```diff
@@ -70,14 +67,15 @@ function makeAccessKey(hostname: string, accessKeyJson: AccessKeyJson): AccessKe
       accessKeyJson.id, accessKeyJson.name, accessKeyJson.metricsId, proxyParams);
 }
 
-function makeAccessKeyJson(accessKey: AccessKey): AccessKeyJson {
+function accessKeyToPersistedJson(accessKey: AccessKey): AccessKeyJson {
```
[the remark this answers is not in the document]
Comment (JonathanDCohen) [reply src/shadowbox/server/server_access_key.ts:70]: The rename is to keep from having two separate `accessKeyToJson` functions, one of which is for the API response and one of which is for creating the persisted format

File: src/shadowbox/server/manager_service.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (JonathanDCohen) [reply src/shadowbox/server/manager_service.ts:324]: The shadowbox access key config is written but not the cipher list config file in the proxy server

File: src/metrics_server/README.md [hunk 21]
```diff
@@ -35,6 +35,7 @@ The metrics server supports two URL paths:
         timestampUtcMs: number,
         dataLimit: {
             enabled: boolean
+            perKeyLimitCount: number
```
[the remark this answers is not in the document]
Comment (JonathanDCohen) [reply src/metrics_server/README.md:38]: I'm not sure the difficulties you're worried about -- could you elaborate?  I figured this way would make more sense since we've called it "dataLimit", not "serverDataLimit"``.

File: src/shadowbox/server/manager_service.ts (hunk 19 above)
Comment (JonathanDCohen) [reply src/shadowbox/server/manager_service.ts:48]: That's I think more technically correct JSON.  It's also consistent with, for example, our response in `getServer()`

File: src/shadowbox/server/server_access_key.ts [hunk 22]
```diff
@@ -212,8 +219,8 @@ export class ServerAccessKeyRepository implements AccessKeyRepository {
     for (const accessKey of this.accessKeys) {
       const usageBytes = bytesTransferredById[accessKey.id] || 0;
       const wasOverDataLimit = accessKey.isOverDataLimit;
-      accessKey.isOverDataLimit =
-          this.accessKeyDataLimit ? usageBytes > this.accessKeyDataLimit.bytes : false;
+      const limitBytes = (accessKey.dataLimit || this.defaultDataLimit)?.bytes;
```
[the remark this answers is not in the document]
Comment (JonathanDCohen) [reply src/shadowbox/server/server_access_key.ts]: Done in 1cf0f6b

File: src/shadowbox/model/access_key.ts [hunk 23]
```diff
@@ -44,6 +44,8 @@ export interface AccessKey {
   readonly proxyParams: ProxyParams;
   // Whether the access key has exceeded the data transfer limit.
   readonly isOverDataLimit: boolean;
+  // The key's current data limit.
```
[the remark this answers is not in the document]
Comment (JonathanDCohen) [reply src/shadowbox/model/access_key.ts]: Done in cd9bc71

File: src/shadowbox/integration_test/test.sh [hunk 24]
```diff
@@ -173,6 +173,35 @@ function cleanup() {
   if [[ "${ACCESS_KEY_JSON}" != *"@${NEW_HOSTNAME}:"* ]]; then
     fail "Hostname for new access keys wasn't changed.  Newly created access key: ${ACCESS_KEY_JSON}"
   fi
+
+  # Verify that we can create default data limits
```
[the remark this answers is not in the document]
Comment (JonathanDCohen) [reply src/shadowbox/integration_test/test.sh]: Done in c935b1d

File: src/shadowbox/server/manager_service.ts (hunk 18 above)
Comment (JonathanDCohen) [reply src/shadowbox/server/manager_service.ts]: Done in 7539a17

File: docker (hunk 16 above)
Comment (JonathanDCohen) [reply docker]: Done in 3360a19

File: src/shadowbox/integration_test/util/docker (hunk 17 above)
Comment (JonathanDCohen) [reply src/shadowbox/integration_test/util/docker]: Done in 3360a19

File: src/metrics_server/README.md (hunk 21 above)
[the remark this answers is not in the document]
Comment (fortuna) [reply src/metrics_server/README.md:38]: Nothing concrete, but in general it's a good idea to not couple things that can be decoupled. The two limits are enabled independently.
I don't feel strongly though.

**Review by fortuna — CHANGES_REQUESTED** [review]

File: src/shadowbox/model/access_key.ts [hunk 25]
```diff
@@ -44,6 +44,8 @@ export interface AccessKey {
   readonly proxyParams: ProxyParams;
   // Whether the access key has exceeded the data transfer limit.
   readonly isOverDataLimit: boolean;
+  // The key's current data limit.  If it exists, it overrides the server default data limit.
+  dataLimit?: DataLimit;
```
[the remark this answers is not in the document]
Comment (JonathanDCohen) [reply src/shadowbox/model/access_key.ts]: Done in 589d1b0

File: src/shadowbox/package.json [hunk 26]
```diff
@@ -1,7 +1,7 @@
 {
   "name": "outline-server",
   "private": true,
-  "version": "1.5.0",
+  "version": "1.5.3",
```
Comment (fortuna) [inline src/shadowbox/package.json:4]: Why 1.5.3?
I think we should revert this.

File: src/shadowbox/server/server_access_key.ts [hunk 27]
```diff
@@ -212,8 +220,11 @@ export class ServerAccessKeyRepository implements AccessKeyRepository {
     for (const accessKey of this.accessKeys) {
       const usageBytes = bytesTransferredById[accessKey.id] || 0;
       const wasOverDataLimit = accessKey.isOverDataLimit;
-      accessKey.isOverDataLimit =
-          this.accessKeyDataLimit ? usageBytes > this.accessKeyDataLimit.bytes : false;
+      let limitBytes = (accessKey.dataLimit || this._defaultDataLimit)?.bytes;
```
Comment (fortuna) [inline src/shadowbox/server/server_access_key.ts:223]: This seems to be the recommended operator: https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-7.html#nullish-coalescing

```suggestion
      let limitBytes = (accessKey.dataLimit ?? this._defaultDataLimit)?.bytes;
```

File: src/shadowbox/server/server_access_key.ts [hunk 28]
```diff
@@ -212,8 +220,11 @@ export class ServerAccessKeyRepository implements AccessKeyRepository {
     for (const accessKey of this.accessKeys) {
       const usageBytes = bytesTransferredById[accessKey.id] || 0;
```
Comment (fortuna) [inline src/shadowbox/server/server_access_key.ts:221]:
```suggestion
      const usageBytes = bytesTransferredById[accessKey.id] ?? 0;
```

For consistency. See below.

File: src/shadowbox/server/server_access_key.ts [hunk 29]
```diff
@@ -212,8 +220,11 @@ export class ServerAccessKeyRepository implements AccessKeyRepository {
...
       const wasOverDataLimit = accessKey.isOverDataLimit;
-      accessKey.isOverDataLimit =
-          this.accessKeyDataLimit ? usageBytes > this.accessKeyDataLimit.bytes : false;
+      let limitBytes = (accessKey.dataLimit || this._defaultDataLimit)?.bytes;
+      if (limitBytes === undefined) {
+        limitBytes = Number.POSITIVE_INFINITY;
+      }
```
Comment (fortuna) [inline src/shadowbox/server/server_access_key.ts:226]: This is something you can do in TS4:
```suggestion
      limitBytes ??= Number.POSITIVE_INFINITY;
```

See https://www.typescriptlang.org/docs/handbook/release-notes/typescript-4-0.html#short-circuiting-assignment-operators

**Review by fortuna — COMMENTED** [review]
Thanks for all the changes. This is looking pretty good.
I like how careful you were with adding tests, and I appreciate aligning the language for the default limit.
I only have some minor comments.

**Review by fortuna — COMMENTED** [review]
Let's try to submit it today. Ping me when you've addressed the comments.

@alalamav Last call for a review :-)

**Review by fortuna — APPROVED** [review]

File: src/shadowbox/package.json (hunk 26 above)
Comment (JonathanDCohen) [reply src/shadowbox/package.json:4]: That's [where master is](https://github.com/Jigsaw-Code/outline-server/blob/master/src/shadowbox/package.json).  I'm not sure why this diff is displaying in this way.

File: src/shadowbox/server/server_access_key.ts (hunk 29 above)
Comment (JonathanDCohen) [reply src/shadowbox/server/server_access_key.ts:226]: We're on 3.8.3.  I'll update our dependencies (including to TS 4) in a separate PR

**Review by fortuna — APPROVED** [review]
Thanks for getting this done! We're almost there!
